### PR TITLE
Options for fake collection

### LINKED ISSUE #755: Add overload with optionsBuilder parameter for CollectionOfFake
Unlike `A.Fake`, `A.CollectionOfFake` doesn't have an overload with an `optionsBuilder` parameter. It's probably an oversight, which could easily be remedied.

### PR DESCRIPTION
**Pull request by thomaslevesque**
fixes #755

### CODE REVIEW

File: src/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs [hunk 1]
```diff
@@ -53,25 +53,33 @@ public object CreateFake(Type typeOfFake, Action<IFakeOptions> optionsBuilder)
...
+                result.Add(this.CreateFake(optionsBuilder));
             }
 
             return result;
         }
 
-        public IList<object> CollectionOfFake(Type typeOfFake, int numberOfFakes)
+        public IList<object> CollectionOfFake(Type typeOfFakes, int numberOfFakes)
         {
-            Guard.AgainstNull(typeOfFake, nameof(typeOfFake));
```
Comment (thomaslevesque) [inline src/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs]: This guard clause was redundant, because we end up calling `CreateFake`, which already has the guard clause.

File: tests/FakeItEasy.Specs/CreationSpecs.cs [hunk 2]
```diff
@@ -177,5 +204,10 @@ protected override IList<T> CreateCollectionOfFake<T>(int numberOfFakes)
         {
             return A.CollectionOfFake(typeof(T), numberOfFakes).Cast<T>().ToList();
         }
+
+        protected override IList<T> CreateCollectionOfFake<T>(int numberOfFakes, Action<IFakeOptions<T>> optionsBuilder)
+        {
+            return A.CollectionOfFake(typeof(T), numberOfFakes, options => optionsBuilder((IFakeOptions<T>)options)).Cast<T>().ToList();
```
Comment (adamralph) [inline tests/FakeItEasy.Specs/CreationSpecs.cs]: Why the wrapped call here? Can't we just pass `optionsBuilder` directly?

Comment (thomaslevesque) [inline tests/FakeItEasy.Specs/CreationSpecs.cs]: The non-generic overload accepts an `Action<IFakeOptions>`, not `Action<IFakeOptions<T>>` There's no inheritance relation between the interfaces, so contravariance doesn't work here.

Comment (thomaslevesque) [inline tests/FakeItEasy.Specs/CreationSpecs.cs]: [This code](https://gist.github.com/thomaslevesque/a4ad1fc87dd03edfb3b8dfc589146156) works, but if you remove the inheritance between the interfaces, it doesn't.

File: src/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs [hunk 3]
```diff
@@ -53,25 +53,33 @@ public object CreateFake(Type typeOfFake, Action<IFakeOptions> optionsBuilder)
...
             {
-                result.Add(this.CreateFake<T>(x => { }));
+                result.Add(this.CreateFake(optionsBuilder));
             }
 
             return result;
         }
 
-        public IList<object> CollectionOfFake(Type typeOfFake, int numberOfFakes)
+        public IList<object> CollectionOfFake(Type typeOfFakes, int numberOfFakes)
```
Comment (adamralph) [inline src/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs]: The parameter in `A` is named `typeOfFake` so I think we should be consistent.

If we'd rather pluralise it in `A` then I think that's OK, but only because we haven't released those overloads yet, as this could be called a breaking change (I'm not sure what our breaking/non-breaking policy is on parameter name changes).

Whatever the decision, I'd prefer this commit to be consistent with `A` and remain singular. This would avoid having this unrelated change here. If we decide to pluralise both here and in `A`, that should be a separate commit.

Comment (thomaslevesque) [inline src/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs]: > The parameter in A is named typeOfFake so I think we should be consistent.

Ah, yes... I changed it so that it would be the same as on the interface. I'll make it singular everywhere.

Comment (adamralph) [inline src/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs]: I see that the existing parameter in `IFakeCreatorFacade` is plural so there is already inconsistency. I think the cleanest way is to remove this change from this commit, since it's unrelated to adding the new overloads, and add another commit which makes everything consistent, whether singular or plural. I have no preference which.

Comment (adamralph) [inline src/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs]: My last comment overlapped with yours. Seems we're in agreement.

File: tests/FakeItEasy.Specs/CreationSpecs.cs (hunk 2 above)
Comment (adamralph) [inline tests/FakeItEasy.Specs/CreationSpecs.cs]: Ah OK, makes sense.

File: src/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs (hunk 3 above)
Comment (thomaslevesque) [inline src/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs]: Do you still want a separate commit? If we make it singular everywhere, the only place on `master` where it needs to be fixed is `IFakeCreationFacade.CollectionOfFake(Type typeOfFakes, int numberOfFakes)`

Comment (adamralph) [inline src/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs]: For a such small change, no, same commit is fine.

**Comment by thomaslevesque** [comment]
All done, I think.

**Comment by adamralph** [comment]
Thanks!

**Comment by thomaslevesque** [comment]
Thanks for the comments :smile:

**Comment by blairconrad** [comment]
Nice. Thank you.
Is it worthwhile adding to the docs? At some point, I'm of the mind that "given the docs for the simpler examples, these things are figureoutable"…

**Comment by thomaslevesque** [comment]
> Is it worthwhile adding to the docs?

I considered it, and thought it wasn't. We have examples of `A.Fake` with options and `A.FakeCollection` without, so people will probably figure out they can pass options when creating a collection of fakes, too. And the Intellisense helps, too.

**Comment by adamralph** [comment]
Agreed. No need to cover every single point in the overload matrix in the docs.
